core/stat/base/slidingwindow_metric.go

```diff
@@ -0,0 +1,150 @@
+	windowLengthInMs uint32
+	sampleCount      uint32
+	intervalInMs     uint32
+	real             *BucketLeapArray
+}
+
+// It must pass the parameter point to the real storage entity
+func NewSlidingWindowMetric(intervalInMs, sampleCount uint32, real *BucketLeapArray) *SlidingWindowMetric {
+	if real == nil || intervalInMs <= 0 || sampleCount <= 0 {
+		panic(fmt.Sprintf("Illegal parameters,intervalInMs=%d,sampleCount=%d,real=%+v.\n", intervalInMs, sampleCount, real))
+	}
+
+	if intervalInMs%sampleCount != 0 {
+		panic(fmt.Sprintf("Invalid parameters, intervalInMs is %d, sampleCount is %d.\n", intervalInMs, sampleCount))
+	}
+	winLengthInMs := intervalInMs / sampleCount
+
+	parentIntervalInMs := real.IntervalInMs()
+	parentWindowLengthInMs := real.WindowLengthInMs()
+
+	if winLengthInMs != parentWindowLengthInMs {
+		panic(fmt.Sprintf("SlidingWindowMetric's WindowLengthInMs(%d) is not equal to real BucketLeapArray's WindowLengthInMs(%d).\n", winLengthInMs, parentWindowLengthInMs))
+	}
+
+	if intervalInMs > parentIntervalInMs {
+		// todo if SlidingWindowMetric's intervalInMs is greater than BucketLeapArray.
+		panic(fmt.Sprintf("The length(%d) of SlidingWindowMetric is greater than parent BucketLeapArray(%d).\n", intervalInMs, parentIntervalInMs))
+	}
+
+	// 10 * 1000 ms == parent
+	if parentIntervalInMs%intervalInMs != 0 {
+		panic(fmt.Sprintf("SlidingWindowMetric's intervalInMs(%d) is not divisible by real BucketLeapArray's intervalInMs(%d).\n", intervalInMs, parentIntervalInMs))
+	}
+
+	return &SlidingWindowMetric{
+		windowLengthInMs: winLengthInMs,
+		sampleCount:      sampleCount,
+		intervalInMs:     intervalInMs,
+		real:             real,
+	}
+}
+
+// Get the [start index, end index) in BucketLeapArray for SlidingWindowMetric in current time.
+func (m *SlidingWindowMetric) getIndexInParent(now uint64) (startIdx, endIdx int) {
+	// parent BucketLeapArray's start time
+	s1 := calculateStartTime(now, m.real.WindowLengthInMs())
+	// current SlidingWindowMetric's start time
+	s2 := calculateStartTime(now, m.windowLengthInMs)
+
+	startIdx = int((s2 - s1) / uint64(m.windowLengthInMs))
+	endIdx = startIdx + int(m.sampleCount)
+	return
+}
+
+func (m *SlidingWindowMetric) count(event base.MetricEvent, start, end int, value []*windowWrap) int64 {
+	ret := int64(0)
+	l := len(value)
+	for i := start; i < end; i++ {
+		if i >= l {
+			logger.Panicf("The index(%d) is outbound of len(%d).\n", i, l)
+			continue
+		}
+		w := value[i]
+		counter, ok := w.value.(metricBucket)
+		if !ok {
+			logger.Errorf("Fail to assert windowWrap value(%+v) to metricBucket.\n", w.value)
+			continue
+		}
+		ret += counter.get(event)
+	}
+	return ret
+}
+
+func (m *SlidingWindowMetric) Total() int64 {
+	return m.Pass() + m.Block()
+}
+
+func (m *SlidingWindowMetric) Pass() int64 {
+	now := util.CurrentTimeMillis()
+	start, end := m.getIndexInParent(now)
+	allBuckets := m.real.Values(now)
+	return m.count(base.MetricEventPass, start, end, allBuckets)
```

Comment:
Using the original index is dangerous (and incorrect) here. The retrieved `allBuckets` might have been filtered (some deprecated buckets are not included), so using the original idx here might cause index out-of-bound error. Maybe providing a time predicate when extracting all data buckets via the `Values` method is better?

Reply:
Had fixed it